history/history-impl/src/main/java/com/duckduckgo/history/impl/HistoryRepository.kt

```diff
@@ -0,0 +1,131 @@
+ * limitations under the License.
+ */
+
+package com.duckduckgo.history.impl
+
+import com.duckduckgo.common.utils.DispatcherProvider
+import com.duckduckgo.history.api.HistoryEntry
+import com.duckduckgo.history.impl.remoteconfig.HistoryFeature
+import com.duckduckgo.history.impl.store.HistoryDao
+import com.duckduckgo.history.impl.store.HistoryDataStore
+import io.reactivex.Single
+import java.time.LocalDateTime
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+interface HistoryRepository {
+    fun getHistoryObservable(): Single<List<HistoryEntry>>
+
+    suspend fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    )
+
+    suspend fun clearHistory()
+
+    fun isHistoryUserEnabled(default: Boolean): Boolean
+
+    fun setHistoryUserEnabled(value: Boolean)
+
+    suspend fun clearEntriesOlderThan(dateTime: LocalDateTime)
+
+    suspend fun hasHistory(): Boolean
+}
+
+class RealHistoryRepository(
+    private val historyDao: HistoryDao,
+    private val dispatcherProvider: DispatcherProvider,
+    private val appCoroutineScope: CoroutineScope,
+    private val historyDataStore: HistoryDataStore,
+    private val historyFeature: HistoryFeature,
+) : HistoryRepository {
+
+    private var cachedHistoryEntries: List<HistoryEntry>? = null
+
+    override fun getHistoryObservable(): Single<List<HistoryEntry>> {
+        return if (cachedHistoryEntries != null) {
+            Single.just(cachedHistoryEntries)
+        } else {
+            Single.create { emitter ->
+                appCoroutineScope.launch(dispatcherProvider.io()) {
+                    try {
+                        emitter.onSuccess(fetchAndCacheHistoryEntries())
+                    } catch (e: Exception) {
+                        emitter.onError(e)
+                    }
+                }
+            }
+        }
+    }
+
+    override suspend fun saveToHistory(
+        url: String,
+        title: String?,
+        query: String?,
+        isSerp: Boolean,
+    ) {
+        withContext(dispatcherProvider.io()) {
+            historyDao.updateOrInsertVisit(
+                url,
+                title ?: "",
+                query,
+                isSerp,
+                LocalDateTime.now(),
+            )
+            fetchAndCacheHistoryEntries()
+        }
+    }
+
+    override suspend fun clearHistory() {
+        withContext(dispatcherProvider.io()) {
+            cachedHistoryEntries = null
+            historyDao.deleteAll()
+            fetchAndCacheHistoryEntries()
+        }
+    }
+
+    override fun isHistoryUserEnabled(default: Boolean): Boolean {
+        return historyDataStore.isHistoryUserEnabled(default)
+    }
+
+    override fun setHistoryUserEnabled(value: Boolean) {
+        historyDataStore.setHistoryUserEnabled(value)
+    }
+
+    private suspend fun fetchAndCacheHistoryEntries(): List<HistoryEntry> {
```

Comment:
On a normal usage, I haven't found issues here. However, I was stressing this a bit calling `saveToHistory` on a loop, and since we execute it every time we add a site to history, I noticed this was blocking inserting new sites until all event of fetchAndCached are processed. Reading from db and blocks other writes, which it's expected. Thinking about this, any new event will override results from previous one. So if we have 10 events waiting, we only need to process 1. thoughts? (not a blocker, just leaving here for consideration)

Reply:
As you said, not a blocker, but I think we should observe the database instead